test/unit/math/rev/mat/functor/idas_system_test.cpp

```diff
@@ -0,0 +1,337 @@
+  size_t n = dae.n();
+  auto yy = dae.nv_yy();
+  auto yp = dae.nv_yp();
+  auto yys = dae.nv_yys();
+  auto yps = dae.nv_yps();
+
+  auto p1 = theta[0];
+  auto p2 = theta[1];
+  auto p3 = theta[2];
+
+  N_Vector res = N_VNew_Serial(n);
+  N_Vector* ress = N_VCloneVectorArray(ns, res);
+  for (size_t is = 0; is < ns; is++) {
+    N_VConst(RCONST(1.0), yys[is]);
+    N_VConst(RCONST(1.0), yps[is]);
+  }
+  auto user_data = dae.to_user_data();
+
+  auto residual = dae.residual();
+  double t = 0.0;
+  EXPECT_EQ(residual(t, yy, yp, res, user_data), 0);
+  auto fval = f(t, yy0, yp0, theta, x_r, x_i, msgs);
+  for (size_t i = 0; i < n; ++i)
+    EXPECT_EQ(NV_Ith_S(res, i), fval[i]);
+
+  auto sensitivity_residual = dae.sensitivity_residual();
+  auto temp1 = N_VNew_Serial(n);
+  auto temp2 = N_VNew_Serial(n);
+  auto temp3 = N_VNew_Serial(n);
+  EXPECT_EQ(sensitivity_residual(ns, t, yy, yp, res, yys, yps, ress, user_data,
+                                 temp1, temp2, temp3),
+            0);
+  Eigen::MatrixXd Jyy(n, n), Jyp(n, n), Jpar(n, ns);
+  Jyy << p1, -p2 * yy0[2], -p2 * yy0[1], -p1, p2 * yy0[2] + 2 * p3 * yy0[1],
+      p2 * yy0[1], 1, 1, 1;
+  Jyp << 1, 0, 0, 0, 1, 0, 0, 0, 0;
+  Jpar *= 0.0;
+  Jpar.col(2 * n) << yy0[0], -yy0[0], 0;
+  Jpar.col(2 * n + 1) << -yy0[1] * yy0[2], yy0[1] * yy0[2], 0;
+  Jpar.col(2 * n + 2) << 0, yy0[1] * yy0[1], 0;
+  Eigen::MatrixXd yys_mat(n, ns), yps_mat(n, ns);
+  for (size_t i = 0; i < n; ++i) {
+    for (size_t j = 0; j < ns; ++j) {
+      yys_mat(i, j) = 1.0;
+      yps_mat(i, j) = 1.0;
+    }
+  }
+
+  // check sensitivity_residual result
+  auto r = Jyy * yys_mat + Jyp * yps_mat + Jpar;
+  for (size_t i = 0; i < n; ++i) {
+    for (size_t j = 0; j < ns; ++j) {
+      EXPECT_EQ(NV_Ith_S(ress[j], i), r(i, j));
+    }
+  }
+}
+
+TEST(IDAS_DAE_SYSTEM, constructor_errors) {
+  using stan::math::idas_forward_system;
+  using stan::math::var;
+  chemical_kinetics f;
+  std::vector<double> yy0{0.5, 0.4, 0.2};
+  std::vector<double> yp0{0.1, 0.2, 0.3};
+  std::vector<double> theta{0.040, 1.0e1, 3.0e-1};
+  const std::vector<int> eq_id{1, 1, 0};
+  auto yy0_var = stan::math::to_var(yy0);
+  auto yp0_var = stan::math::to_var(yp0);
+  auto theta_var = stan::math::to_var(theta);
+  std::vector<double> x_r(3, 1);
+  std::vector<int> x_i;
+  std::ostream* msgs = 0;
+
+  std::vector<double> bad_double{yy0};
+  bad_double[0] = std::numeric_limits<double>::infinity();
+  auto build_double = [&f, msgs, &x_i](const std::vector<int>& eq_id,
+                                       const std::vector<double>& yy0,
+                                       const std::vector<double>& yp0,
+                                       const std::vector<double>& theta,
+                                       const std::vector<double>& x_r) {
+    idas_forward_system<chemical_kinetics, double, double, double> dae{
+        f, eq_id, yy0, yp0, theta, x_r, x_i, msgs};
+  };
+
+  ASSERT_NO_THROW(build_double(eq_id, yy0, yp0, theta, x_r));
```

Comment:
Add tests to verify things throw when length(yy0), length(yp0), and length(eq_id) aren't all equal

Reply:
added test